Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,686 @@
+    anything though...).
+    """
+
+    hash_entry = node.scope.lookup_here("__hash__")
+    if hash_entry:
+        # TODO ideally assignment of __hash__ to None shouldn't trigger this
+        # but difficult to get the right information here
+        if unsafe_hash:
+            # error message taken from CPython dataclasses module
+            error(node.pos, "Cannot overwrite attribute __hash__ in class %s" % node.class_name)
+        return "", {}, []
+    if not unsafe_hash:
+        if eq and not frozen:
+            return "", {}, [ Nodes.SingleAssignmentNode(node.pos,
+                             lhs = ExprNodes.NameNode(node.pos, name=EncodedString("__hash__")),
+                             rhs = ExprNodes.NoneNode(node.pos)) ]
+        if not eq:
+            return
+
+    names = [
+        name for name, field in fields.items()
+        if (not field.is_initvar and
+            (field.compare.value if field.hash.value is None else field.hash.value))
+    ]
+    if not names:
+        return "", {}, []  # nothing to hash
+
+    # make a tuple of the hashes
+    tpl = u", ".join(u"hash(self.%s)" % name for name in names )
+
+    # if we're here we want to generate a hash
+    code_lines = dedent(u"""\
+        def __hash__(self):
+            return hash((%s))
+        """) % tpl
+
+    return code_lines, {}, []
+
+
+class GetTypeNode(ExprNodes.ExprNode):
+    # Tries to return a pytype_type if possible. However contains
+    # some fallback provision if it turns out not to resolve to a Python object.
+    # Initialize with "entry".
+
+    subexprs = []
+
+    def __init__(self, entry):
+        super(GetTypeNode, self).__init__(entry.pos, entry=entry)
+
+    def analyse_types(self, env):
+        type = self.entry.type
+
+        if type.is_extension_type or type.is_builtin_type:
+            return ExprNodes.RawCNameExprNode(
+                self.pos, Builtin.type_type, type.typeptr_cname).analyse_types(env)
+        else:
+            names = None
+            py_name = type.py_type_name()
+            # int types can return "(int, long)"
+            if py_name:
+                names = [ name.strip("() ") for name in py_name.split(",") ]
+            if names:
+                for name in names:
+                    name = EncodedString(name)
+                    nn = ExprNodes.NameNode(self.pos, name=name)
+                    nn = nn.analyse_types(env)
+                    if nn.entry:
+                        return nn
+
+        # otherwise we're left to return a string
+        s = self.entry.annotation.string.value if self.entry.annotation else None
+        if not s:
+            s = self.entry.type.declaration_code("", for_display=1)
+        return ExprNodes.StringNode(self.pos, value=s).analyse_types(env)
+
+
+class FieldsValueNode(ExprNodes.ExprNode):
+    # largely just forwards arg. Allows it to be coerced to a Python object
+    # if possible, and if not then generates a sensible backup string
```

Comment:
Could this not be tidied up and used as a doc-string? (Do the same for GetTypeNode).